app/src/main/java/io/appium/settings/RecorderService.java

```diff
@@ -0,0 +1,138 @@
+                    (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
+
+            startScreenRecord(mMediaProjectionManager, intent);
+        } else if (ACTION_RECORDING_STOP.equals(action)) {
+            Log.v(TAG, "onStartCommand:intent=" + "stop");
+            stopScreenRecord();
+            result = START_NOT_STICKY;
+        }
+
+        return result;
+    }
+
+    /**
+     * start screen recording
+     */
+    @SuppressLint("NewApi")
+    private void startScreenRecord(MediaProjectionManager mediaProjectionManager,
+                                   final Intent intent) {
+        if (recorderThread == null) {
+            Log.v(TAG, "recorder thread null");
+
+            final int resultCode = intent.getIntExtra(ACTION_RECORDING_RESULT_CODE, 0);
+            // get MediaProjection
+            final MediaProjection projection = mediaProjectionManager.getMediaProjection(resultCode,
+                    intent);
+            if (projection != null) {
+                Log.v(TAG, "projection different from null");
+                final DisplayMetrics metrics = getResources().getDisplayMetrics();
+                int width = metrics.widthPixels;
+                int height = metrics.heightPixels;
+                if (width > height) {
+                    // Horizontal
+                    final float scale_x = width / 1920f;
+                    final float scale_y = height / 1080f;
+                    final float scale = Math.max(scale_x,  scale_y);
+                    width = (int)(width / scale);
+                    height = (int)(height / scale);
+                } else {
+                    // Vertical
+                    final float scale_x = width / 1080f;
+                    final float scale_y = height / 1920f;
+                    final float scale = Math.max(scale_x,  scale_y);
+                    width = (int)(width / scale);
+                    height = (int)(height / scale);
+                }
+                Log.v(TAG, String.format("startRecording:(%d,%d)(%d,%d)",
+                        metrics.widthPixels, metrics.heightPixels, width, height));
+                String outputFilePath = intent.getStringExtra(ACTION_RECORDING_FILENAME);
+                if (outputFilePath != null) {
+                    recorderThread = new RecorderThread(projection, outputFilePath, width, height,
+                            calcBitRate(30, width, height));
+                    recorderThread.startRecording();
+                }
+            }
+        }
+    }
+
+    protected int calcBitRate(final int frameRate, int width, int height) {
+        final int bitrate = (int)(0.25f * frameRate * width * height);
+        Log.i(TAG, String.format("bitrate=%5.2f[Mbps]", bitrate / 1024f / 1024f));
+        return bitrate;
+    }
+
+    /**
+     * stop screen recording
+     */
+    private void stopScreenRecord() {
+        if (recorderThread != null) {
+            recorderThread.stopRecording();
```

Comment:
should we assign recorderThread to null?

Reply:
done